src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala

```diff
@@ -0,0 +1,129 @@
+
+        val ast = checkedTable.ast
+        val mainContractOption = findMainContract(ast)
+        if (mainContractOption.isEmpty) {
+            throw new RuntimeException("No main contract found")
+        }
+
+        val mainName = findMainContractName(ast)
+
+        val finalOutputPath = options.outputPath match {
+            case Some(p) =>
+                Paths.get(p).resolve(mainName)
+            case None =>
+                Paths.get(mainName)
+        }
+
+        // temporary code, not designed to be recursive
+        val mainContract = mainContractOption.get
+        mainContract match {
+            case obsContract: ObsidianContractImpl => {
+                val translated_obj =  translateContract(obsContract)
+                yulString(translated_obj, finalOutputPath)
+            }
+            case _ => None
+        }
+//        val ast_seq =  translateProgram(ast)
+        return true
+    }
+
+    def translateProgram(program: Program): Unit = {
+        // ignore imports for now
+        for (c <- program.contracts) {
+            c match {
+                case obsContract: ObsidianContractImpl => translateContract(obsContract)
+                case javaContract: JavaFFIContractImpl => None
+            }
+        }
+    }
+
+    def translateContract(contract: ObsidianContractImpl): Object = {
+        // create runtime obj
+        var subobjects: Seq[Object] = Seq()
+        val runtime_name = contract.name + "_deployed"
+        val runtime_obj = Object(runtime_name, Code(Block(Seq())), Seq(), Seq())
+        subobjects = runtime_obj +: subobjects
+
+        var statement_seq: Seq[YulStatement] = Seq()
+        // initialization
+//        statement_seq = statement_seq +: translate_mem_init() // TODO check seq append
+
+        // declaration
+        // what to do when instance of constructor
+        for (decl <- contract.declarations if !decl.isInstanceOf[Constructor]) {
+//            statement_seq = statement_seq +: translateDeclaration(decl)
+        }
+
+        val create = Object(contract.name, Code(Block(statement_seq)), subobjects, Seq())
+        return create
+    }
+
+//    def translate_mem_init(): Seq[Statement]
+
+    def translateDeclaration(declaration: Declaration): Seq[YulStatement] = {
+        declaration match {
+            case f: Field => translateField(f)
+            case t: Transaction =>Seq()
+            case s: State =>Seq()
+            case c: ObsidianContractImpl =>Seq()
+            case c: JavaFFIContractImpl => Seq()
+            case c: Constructor =>Seq()
+            case t: TypeDecl =>
+                assert(false, "TODO")
+                Seq()
+            // This should never be hit.
+            case _ =>
+                assert(false, "Translating unexpected declaration: " + declaration)
+                Seq()
+        }
+    }
+
+    def translateField(field: Field): Seq[YulStatement] = {
+        // no need to sstore
```

Comment:
why no need to sstore?

Reply:
oh because the field declaration does not contain actual assignment? Correct me if I am wrong

Reply:
I meant that you should elaborate in the comment.

Reply:
ohhhhhh gotcha